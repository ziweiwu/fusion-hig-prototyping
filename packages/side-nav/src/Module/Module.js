import React, { Component } from "react";
import PropTypes from "prop-types";
import ModulePresenter from "./presenters/ModulePresenter";

export default class Module extends Component {
  static propTypes = {
    /** Indicates this module is currently active */
    active: PropTypes.bool,
    /** Indicates a nested submodule is currently active */
    activeChildren: PropTypes.bool,
    /** Zero or more SideNav submodules */
    children: PropTypes.node,
    /** An instance of @hig/icon */
    icon: PropTypes.node,
    /** URL to navigate to when clicking this module */
    link: PropTypes.string,
    /** Indicates whether to display nested submodules */
    minimized: PropTypes.bool,
    /** Called when clicking on the collapse button */
    onClickCollapseButton: PropTypes.func,
    /** Called when clicking on the title */
    onClickTitle: PropTypes.func,
    /** Called when hovering over the title */
    onMouseOver: PropTypes.func,
    /** A label for rendering this Module */
    title: PropTypes.string.isRequired,
    /** Anchor target. Applicable only if link is provided */
    target: PropTypes.oneOf(["_self", "_blank", "_parent", "_top"])
  };

  static defaultProps = {
    minimized: false
  };

  constructor(props) {
    super(props);

    this.state = {
      minimized: props.minimized
    };
  }

  /**
   * @param {Function} callback, function to call after minimize state has been toggled
   * @returns {Function}
   */
  createMinimizeToggler = callback => () => {
    this.setState(
      { minimized: !this.state.minimized },
      () => callback && callback()
    );
  };

  /**
   * @returns {PresenterBag}
   */
  createPresenterBag() {
    const { minimized } = this.state;

    return {
      minimized,
      onClickCollapseButton: this.createMinimizeToggler(
        this.props.onClickCollapseButton
      ),
      onClickTitle: this.createMinimizeToggler(this.props.onClickTitle)
    };
  }

  /**
   * @param {PresenterBag}
   */
  renderPresenter = presenterBag => {
    const presenterProps = { ...this.props, ...presenterBag };
    const { children, ...otherProps } = presenterProps;

    return <ModulePresenter {...otherProps}>{children}</ModulePresenter>;
  };

  render() {
    const presenterBag = this.createPresenterBag();
    const { renderPresenter } = this;

    return renderPresenter(presenterBag);
  }
}
