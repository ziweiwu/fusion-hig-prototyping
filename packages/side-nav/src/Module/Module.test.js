import React from "react";
import renderer from "react-test-renderer";
import { mount } from "enzyme";

import Module from "./Module";

describe("side-nav/Module", () => {
  describe("onClickTitle", () => {
    it("adds click handler to title", () => {
      const onClickTitle = jest.fn();
      const wrapper = mount(
        <Module title="Module" onClickTitle={onClickTitle} />
      );

      wrapper.find(".hig__side-nav__module__link__title").simulate("click");
      expect(onClickTitle).toHaveBeenCalled();
    });
  });

  describe("onClickCollapseButton", () => {
    it("adds click handler to collapse button", () => {
      const onClickCollapseButton = jest.fn();
      const wrapper = mount(
        <Module title="Module" onClickCollapseButton={onClickCollapseButton}>
          Content
        </Module>
      );

      wrapper.find(".hig__side-nav__module__collapse-button").simulate("click");
      expect(onClickCollapseButton).toHaveBeenCalled();
    });
  });

  describe("snapshot tests", () => {
    const cases = [
      {
        description: "renders with mionimal props",
        props: {
          title: "Module"
        }
      }
    ];

    cases.forEach(({ description, props: { children, ...otherProps } }) => {
      it(description, () => {
        const wrapper = <Module {...otherProps}>{children}</Module>;
        const tree = renderer.create(wrapper).toJSON();

        expect(tree).toMatchSnapshot();
      });
    });
  });
});
