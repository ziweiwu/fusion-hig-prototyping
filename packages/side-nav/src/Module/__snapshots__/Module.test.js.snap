// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`side-nav/Module snapshot tests renders with mionimal props 1`] = `
<div
  className="hig--light-theme hig__side-nav__module"
  onMouseOver={[Function]}
>
  <div
    className="hig__side-nav__module__row"
  >
    <div
      className="hig--light-theme hig__side-nav__module__link"
      href={undefined}
      onClick={[Function]}
      target="_self"
    >
      <div
        className="hig__side-nav__module__link__icon"
      />
      <div
        className="hig__side-nav__module__link__title"
      >
        Module
      </div>
      <div
        className="hig__side-nav__module__link__external-link-icon"
      />
    </div>
  </div>
  <div
    className="hig__side-nav__module__submodules hig__side-nav__module__submodules--no-icon"
  />
</div>
`;
